crates/holochain/src/conductor/interface/websocket.rs

```diff
@@ -171,11 +183,28 @@ async fn handle_shutdown(listener_handles: Vec<JoinHandle<InterfaceResult<()>>>)
 
 /// Polls for messages coming in from the external client.
 /// Used by Admin interface.
-async fn recv_incoming_admin_msgs<A: InterfaceApi>(api: A, mut rx_from_iface: WebsocketReceiver) {
+async fn recv_incoming_admin_msgs<A: InterfaceApi>(
+    api: A,
+    mut rx_from_iface: WebsocketReceiver,
+    mut tx_to_iface: WebsocketSender,
+    num_connections: Arc<AtomicUsize>,
+) {
     while let Some(msg) = rx_from_iface.next().await {
         match handle_incoming_message(msg, api.clone()).await {
-            Err(InterfaceError::Closed) => break,
-            Err(e) => error!(error = &e as &dyn std::error::Error),
+            Err(InterfaceError::Closed) => {
+                if let Err(e) =
+                    WebsocketSender::close(&mut tx_to_iface, 1000, "Shutting down".into()).await
+                {
+                    warn!("Admin socket failed to close: {}", e);
+                }
+                if num_connections.load(Ordering::SeqCst) > 0 {
+                    num_connections.fetch_sub(1, Ordering::SeqCst);
```

Comment:
Alas, this is not atomic... it could change between these operations

Reply:
Is it even possible for the connection count to go negative?

Even if so, this non-atomic check can't hurt, it seems like a sanity check that might fail in an extreme race condition. But I wonder if it's necessary at all.

Reply:
yeh how are you meant to do checked subs with atomics?

Reply:
I think this is better now